### ISSUE TITLE
Crash of bspwm

### ISSUE THREAD

**Issue by msteen** [baskerville/bspwm#683]
So far `bspwm` has crashed at least three times. I say crash, but it did not segfault or anything, it just became unresponsive. Each time this happened, luckily all the applications I had running were crash proof (e.g. auto save and restore session), but I obviously would want to see this fixed. I checked the journal, but each time there were either irrelevant or even no entries at all at the time of the crash. This leaves me with three questions:

1. Do you have any clue what might cause the crash? Are there still open issues related to crashes? The mouse still moved, but the cursor state was fixed (i.e. the cursor icon) and I was unable to click anything.  I forgot to try `DISPLAY=:0 bspc ...` in one of my TTYs, that could have given me a more definite answer whether `bspwm` fully crashed. My keyboard shortcuts stopped working as well, I will add a command that does nothing WM related so I can know for sure it did crash as well. One time I was only able to interact with the terminal I had open at the time. This might very well be a different crash, or in the other cases I just happened to have focused something which did not react to keyboard input. I doubt this is enough information to determine the cause of, hence my second question.

2. How can I help pinpoint the cause, if such a crash might happen again in the future?

3. This would depend on the nature of the crash, but if the crash in anyway is known to `bspwm`, would it be possible to dump the state of the window manager, or something similar, such that `bspwm` can be restored fully functional when using the loop example. I ask this, because the last time it crashed I tried `pkill bspwm` and it succesfully restarted `bspwm` with all my windows in tact, but because the window manager state was lost I still had to restart.

**Comment by baskerville** [comment]
1. The first thing that comes to mind is that, if you use `subscribe --fifo`, the initial call to `fopen` will block on the server if the FIFO has no readers. But that would only account for the freezing scenario.

2. You can build a debug version with `make debug`. And then, if it's really a crash, you might interact with a core dump via `systemd-coredumpctl`. Or if *bspwm* is still running you could attach to it (from a VT) with `gdb bspwm $(pgrep -x bspwm)` and see what it might be doing with `bt full`.

3. That's a good idea and I'll try to come up with something.

**Comment by msteen** [comment]
1. I have yet to update to the latest version that contains the fifo additions, so it cannot be that. I will follow your pointers and see if I can give your more to go on.

2. Thanks for the pointers! I will rebuild `bspwm` in debug mode and will update `bspwm` to the latest version just in case (my last update was the stdout exit should exit `bspc subscribe` fix).

3. That would be great! It would change these kind of crashes from being a pain to a mere annoyance. Obviously crash free is the ideal that should be pursued, but crash resistance is a great property to have in those situations where this is unfortunately still not the case.

**Comment by msteen** [comment]
I just had another crash, unfortunately I had yet to recompile `bspwm` in debug mode, so I could not pinpoint the problem, but I think it might be caused by having a `synergy` server run in the background. I tried `DISPLAY=:0 bspc query -N -n` in one of my TTYs and it just blocked without any response. And since restarting `bspwm` solves the problem, it should not be X that causes the block, right? I will recompile `bspwm` now and will make sure I have `synergy` server running in the background to see if I can reproduce the bug in debug mode.

**Comment by baskerville** [comment]
3. In the meantime you might use the following:
```sh
if [ -e "$BSPWM_STATE" ] ; then
	bspc wm -l "$BSPWM_STATE"
	rm "$BSPWM_STATE"
else
	bspc wm -o
fi
```

**Comment by msteen** [comment]
It finally crashed again. That I happened to be running `synergy` at the time probably was just a  coincidence. The things of note are.

- The crash occurred when I clicked a Google Chrome window.
- The CPU usage of `X` has been higher than it should be, but this was the case even before crashing, so this likely is unrelated.
- I log filled up with `ERROR:render_media_log.cc(30) MediaEvent: MEDIA_ERROR_LOG_ENTRY {"error": "Large timestamp gap detected: may cause AV sync to drift. time expected delta"}`. This is probably caused by a new media key extension I am using in Chrome. Even though the crash occurred when clicking Chrome, I did not see these errors after previous crashes, so it probably is unrelated.
- The backtrace gives me the following: 

      __write_no_cancel
      _IO_new_file_write
      new_do_write
      __GI__IO_do_write
      __GI_IO_file_sync
      fflush
      print_report
      put_status
      focus_node
      grab_pointer
      button_press
      main

Given this backtrace, am I correct in my understanding that this `fflush` call that has to succeed before it can continue, but for some reason the stream to which it tries to flush is unable to heed this request, causing `bspwm` main thread to block entirely?

And given the call to `print_report` it has to be my panel setup that causes the problem. Unfortunately I do not know what might cause the flush to fail, since it is not as if I just (re)started my panel.

If my above understanding is correct, the solution would likely have to be to have some crash policy in place for these kind of situations, making sure that the main thread can never block due to them. And with regards to the actual cause, I will have to take another look at my `lemonbar` setup.

**Comment by msteen** [comment]
Considering that the cause of the high CPU usage of `X` and the terminal I was using, was caused by a memory leak in the terminal, it might very well have been the cause of the crash. I also updated my lemonbar script to only open the fifo with write mode, maybe having it in both read/write mode caused a deadlock of sorts, not sure, could be another reason for the `fflush` blocking. Either way it seems external, so I am closing the issue for now.
